.travis.yml

```diff
@@ -1,16 +1,39 @@
 sudo: false
-
 language: python
+cache: pip
 
-python:
-  - "3.6"
-  - "pypy3"
-
-install:
-  - pip install tox-travis codecov
-
-script:
-  - tox
+stages:
+  - test
+  - name: build container
+    if: branch != master AND tag IS blank
+  - name: build and push container
+    if: (branch = master) OR (tag IS present)
 
-after_success:
-  - codecov
+jobs:
+  include:
+    - stage: test
+      python: 3.6
+      install: pip install tox-travis codecov
+      script: tox
+      after_success: codecov
+    - stage: test
+      python: pypy3
+      install: pip install tox-travis codecov
+      script: tox
+      after_success: codecov
+    - stage: build container
+      services: docker
+      sudo: required
+      script:
+        - docker build .
+    - stage: build and push container
+      services: docker
+      sudo: required
+      script:
+        - docker login -u "$DOCKER_USERNAME" -p "$DOCKER_PASSWORD"
```

Comment:
Where are `DOCKER_USERNAME` and `DOCKER_PASSWORD` set? I presume there's some magic environment thing I'm just not seeing. Do each of us need to add these to our Travis-CI environment settings (or wherever applicable)?

Reply:
Sorry, I should have clarified that. They are set in the Travis dashboard and belong to a special user I created for pushing docker images. 

There is no need to set them anywhere, as they'll be pulled for our builds automagically.